js-extensions/src/ExtensionStore.js

```diff
@@ -0,0 +1,316 @@
+            ResourceLoadTracker.setExtensionPointMetadata(this.extensionPointList);
+            if (oncomplete) oncomplete(this.extensionPointList);
+        });
+    }
+
+    /**
+     * Load the bundles for the given type
+     */
+    _loadBundles(extensionPointId, onload) {
+        // Make sure this has been initialized first
+        if (!this.extensionPointList) {
+            this._loadExtensionData(() => {
+                this._loadBundles(extensionPointId, onload);
+            });
+            return;
+        }
+        
+        var extensionPointMetadatas = this.extensionPoints[extensionPointId];
+        if (extensionPointMetadatas && extensionPointMetadatas.loaded) {
+            onload(extensionPointMetadatas);
+            return;
+        }
+        
+        extensionPointMetadatas = this.extensionPoints[extensionPointId] = this.extensionPoints[extensionPointId] || [];
+        extensionPointMetadatas.loaded = true;
+        
+        var jsModules = require('@jenkins-cd/js-modules');
+        var loadCountMonitor = new LoadCountMonitor();
+        
+        var loadPluginBundle = (pluginMetadata) => {
+            loadCountMonitor.inc();
+            
+            // The plugin bundle for this plugin may already be in the process of loading (async extension
+            // point rendering). If it's not, pluginMetadata.loadCountMonitors will not be undefined,
+            // which means we can go ahead with the async loading. If it is, pluginMetadata.loadCountMonitors
+            // is defined, we just add "this" loadCountMonitor to pluginMetadata.loadCountMonitors.
+            // It will get called as soon as the script loading is complete.
+            if (!pluginMetadata.loadCountMonitors) {
+                pluginMetadata.loadCountMonitors = [];
+                pluginMetadata.loadCountMonitors.push(loadCountMonitor);
+                jsModules.import(pluginMetadata.hpiPluginId + ':jenkins-js-extension')
+                    .onFulfilled(() => {
+                        pluginMetadata.bundleLoaded = true;
+                        for (var i = 0; i < pluginMetadata.loadCountMonitors.length; i++) {
+                            pluginMetadata.loadCountMonitors[i].dec();
+                        }
+                        delete pluginMetadata.loadCountMonitors;
+                    });
+            } else {
+                pluginMetadata.loadCountMonitors.push(loadCountMonitor);
+            }
+        };
+        
+        var checkLoading = () => {
+            if (loadCountMonitor.counter === 0) {
+                onload(extensionPointMetadatas);
+            }
+        };
+    
+        // Iterate over each plugin in extensionPointMetadata, async loading
+        // the extension point .js bundle (if not already loaded) for each of the
+        // plugins that implement the specified extensionPointId.
+        for(var i1 = 0; i1 < this.extensionPointList.length; i1++) {
```

Comment:
yes, agree, but we use es6 in all the rest of the PR and IMO we should be consistent with it (and all the good reasons you linked ;))